### PR TITLE
Redux feature

### PR DESCRIPTION
**Pull request by pavliukvlad**

### CODE REVIEW

File: StudentList/Common/Extensions/StudentFilterExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+                && studentFilter.Group == StudentFilter.Default.Group
+                && studentFilter.Birthdate == StudentFilter.Default.Birthdate)
+            {
+                return true;
+            }
+
+            return false;
+        }
+    }
+}
```
Comment (igor-leshkevych) [inline StudentList/Common/Extensions/StudentFilterExtensions.cs]: Why extension method and not class method?

File: StudentList/Common/Extensions/StudentFilterExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public static class StudentFilterExtensions
+    {
+        public static bool IsDefault(this StudentFilter studentFilter)
+        {
+            if (studentFilter.Name == StudentFilter.Default.Name
+                && studentFilter.Group == StudentFilter.Default.Group
+                && studentFilter.Birthdate == StudentFilter.Default.Birthdate)
+            {
+                return true;
```
Comment (igor-leshkevych) [inline StudentList/Common/Extensions/StudentFilterExtensions.cs]: can be written as `return studentFilter.Name == StudentFilter.Default.Name
                && studentFilter.Group == StudentFilter.Default.Group
                && studentFilter.Birthdate == StudentFilter.Default.Birthdate`

File: StudentList/Domain/Actions/StudentReceived.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+using StudentList.Models;
+
+namespace StudentList.Domain.Actions
+{
+    public class StudentReceived
```
Comment (igor-leshkevych) [inline StudentList/Domain/Actions/StudentReceived.cs]: studentS

File: StudentList/Domain/Actions/ThunkActions/EditStudentAction.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            this.StudentID = studentId;
+            this.Action = this.EditStudent;
+        }
+
+        public string StudentID { get; }
+
+        public ThunkAction<ApplicationState> Action { get; set; }
+
+        private void EditStudent(Dispatcher dispatcher, Func<ApplicationState> getState)
```
Comment (igor-leshkevych) [inline StudentList/Domain/Actions/ThunkActions/EditStudentAction.cs]: Why not make this public and remove Action property?

File: StudentList/Domain/Actions/ThunkActions/EditStudentAction.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            var selectedStudent = getState().StudentList.Where(student => student.Id == this.StudentID).FirstOrDefault();
```
Comment (igor-leshkevych) [inline StudentList/Domain/Actions/ThunkActions/EditStudentAction.cs]: `getState().StudentList.FirstOrDefault(student => student.Id == this.StudentID);`. And also StudentID must be StudentId

File: StudentList/Domain/Reducers/StudentListReducer.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+                case StudentListChanged studentListUpdated:
+                    return null;
+                default:
+                    return state;
+            }
+        }
+
+        private static IEnumerable<Student> Reduce(StudentReceived studentReceivedAction)
+        {
+            return studentReceivedAction.StudentList;
```
Comment (igor-leshkevych) [inline StudentList/Domain/Reducers/StudentListReducer.cs]: You should always project results before saving to state, so you ensure that in state you have NEW objects and not same objects as in action. Also, Student in state must be immutable, and you just reuse your Mutable student class, fix it.

File: StudentList/Domain/Reducers/StudentProfileReducer.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+                case AddNewStudent addNewStudent:
+                    return Reduce(addNewStudent);
+                default:
+                    return state;
+            }
+        }
+
+        private static StudentProfileState Reduce(AddNewStudent addNewStudent)
+        {
+            return new StudentProfileState(null);
```
Comment (igor-leshkevych) [inline StudentList/Domain/Reducers/StudentProfileReducer.cs]: This shouldn't be null, it must be empty student object.

File: StudentList/Domain/States/ApplicationState.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+using StudentList.Models;
+
+namespace StudentList.Domain.States
+{
+    public class ApplicationState
+    {
+        public ApplicationState()
+            : this(
+                filterStudentsState: new FilterStudentsState(StudentFilter.Default),
+                studentList: null,
```
Comment (igor-leshkevych) [inline StudentList/Domain/States/ApplicationState.cs]: Avoid using nulls for List properties or fields, they must be empty lists.

File: StudentList/Domain/Actions/StudentListChanged.cs [hunk 9]
```diff
@@ -0,0 +1,6 @@
+﻿namespace StudentList.Domain.Actions
+{
+    public class StudentListChanged
```
Comment (igor-leshkevych) [inline StudentList/Domain/Actions/StudentListChanged.cs:3]: What this action is supposed to indicate?

File: StudentList/Domain/Actions/AddNewStudent.cs [hunk 10]
```diff
@@ -0,0 +1,6 @@
+﻿namespace StudentList.Domain.Actions
+{
+    public class AddNewStudent
+    {
```
Comment (igor-leshkevych) [inline StudentList/Domain/Actions/AddNewStudent.cs]: Why this action is without payload?

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 11]
```diff
@@ -87,29 +92,24 @@ public override View OnCreateView(LayoutInflater inflater, ViewGroup container,
...
-            this.birthdateLayout.EditText.Text = string.IsNullOrWhiteSpace(this.StudentId) ? string.Empty
+            this.nameLayout.EditText.Text = this.selectedStudent == null ? string.Empty : this.selectedStudent.Name;
+            this.birthdateLayout.EditText.Text = this.selectedStudent == null ? string.Empty
                 : this.selectedStudent.Birthdate.ToString(FormatConstants.DateTimeFormat, CultureInfo.InvariantCulture);
-            this.universityLayout.EditText.Text = string.IsNullOrWhiteSpace(this.StudentId) ? string.Empty : this.selectedStudent.University;
-            this.groupLayout.EditText.Text = string.IsNullOrWhiteSpace(this.StudentId) ? string.Empty : this.selectedStudent.GroupName;
-            this.phoneLayout.EditText.Text = string.IsNullOrWhiteSpace(this.StudentId) ? string.Empty : this.selectedStudent.Phone;
-            this.confirmButton.Text = this.StudentId == null ? this.GetString(Resource.String.add_new_student_btn)
+            this.universityLayout.EditText.Text = this.selectedStudent == null ? string.Empty : this.selectedStudent.University;
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs]: `this.selectedStudent?.University ?? string.Empty`

File: StudentList/Fragments/StudentListFragment.cs [hunk 12]
```diff
@@ -117,18 +132,51 @@ public override bool OnOptionsItemSelected(IMenuItem item)
...
-            this.studentsCountTextView.Text = string.Format(
-               CultureInfo.InvariantCulture, this.GetString(Resource.String.student_count_pattern), this.studentAdapter.ItemCount);
+        private void HandleNewStudents(StudentListModel model)
+        {
+            if (model.Students != null)
+            {
+                var filters = model.StudentFilter;
+                var students = model.Students;
+
+                if (!filters.IsDefault())
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentListFragment.cs]: Extract filtering to a separate function

File: StudentList/Fragments/StudentListFragment.cs [hunk 13]
```diff
@@ -117,18 +132,51 @@ public override bool OnOptionsItemSelected(IMenuItem item)
 
         private void StudentAdapterItemClick(object sender, Student e)
         {
-            this.ShowStudentInfo(e.Id);
+            var editStudentAction = new EditStudentAction(e.Id);
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentListFragment.cs]: `StudentEditRequested` is better name for this action, `EditStudentAction` is better for action that actually commits changes (when you press 'Confirm' on edit student screen).

File: StudentList/Fragments/StudentListFragment.cs [hunk 14]
```diff
@@ -81,13 +89,18 @@ public override void OnStart()
             base.OnStart();
 
             this.studentAdapter.ItemClick += this.StudentAdapterItemClick;
+            this.subsription = this.store.Select(
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentListFragment.cs]: `subscription` name is too general, later if you have more of them you will need to rename it.

File: StudentList/Fragments/StudentListFragment.cs [hunk 15]
```diff
@@ -58,19 +69,16 @@ public override View OnCreateView(LayoutInflater inflater, ViewGroup container,
...
         {
-            if (!this.studentAdapter.IsAnyStudents)
+            if (this.store.GetState().StudentList == null)
             {
-                var students = await this.Activity.RunMethodWithLoaderAsync(
-                    this.repository.GetStudentsAsync(this.studentFilter));
-                this.studentAdapter.SetItems(students);
+                var repoStudents = await this.Activity.RunMethodWithLoaderAsync(
+                    this.repository.GetStudentsAsync(StudentFilter.Default));
+                this.store.Dispatch(new StudentReceived() { StudentList = repoStudents });
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentListFragment.cs]: This should be done in thunk action, not in fragment

**Review by igor-leshkevych — CHANGES_REQUESTED** [review]

File: StudentList/Domain/Actions/StudentListChanged.cs (hunk 9 above)
Comment (pavliukvlad) [reply StudentList/Domain/Actions/StudentListChanged.cs:3]: When I add new student or change existing, I handle this action and return null student's collection so it means app should load students from the repository

**Comment by igor-leshkevych** [comment]
Nice timings @vladpavlyuk 👍

**Comment by pavliukvlad** [comment]
@igor-leshkevych it was really hard task

**Pull request merged**